src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -1275,35 +1301,37 @@ bool MatchesConstructorParameter(ParameterGenerationSpec paramSpec)
                         break;
                     case IFieldSymbol fieldInfo:
                         {
-                            isPublic = fieldInfo.DeclaredAccessibility is Accessibility.Public;
                             isReadOnly = fieldInfo.IsReadOnly;
 #if ROSLYN4_4_OR_GREATER
                             isRequired = fieldInfo.IsRequired;
 #endif
-                            if (fieldInfo.DeclaredAccessibility is not (Accessibility.Private or Accessibility.Protected))
+                            if (fieldInfo.DeclaredAccessibility is Accessibility.Public)
                             {
+                                isPublic = true;
                                 canUseGetter = true;
                                 canUseSetter = !isReadOnly;
                             }
+                            else
+                            {
+                                // Unlike properties JsonIncludeAttribute is not supported for internal fields.
+                                isJsonIncludeInaccessible = hasJsonInclude;
+                            }
                         }
                         break;
                     default:
-                        throw new InvalidOperationException();
+                        Debug.Fail("Method given an invalid symbol type.");
+                        break;
                 }
             }
 
-            private static bool PropertyAccessorCanBeReferenced(MethodInfo? accessor)
-                => accessor != null && (accessor.IsPublic || accessor.IsAssembly);
-
-            private TypeRef? GetConverterTypeFromAttribute(AttributeData attributeData)
+            private TypeRef? GetConverterTypeFromAttribute(INamedTypeSymbol contextType, AttributeData attributeData)
```

Comment:
Re https://github.com/dotnet/runtime/issues/81702#issuecomment-1577321673: we don't appear to issue a diagnostic when there's a the attribute specifies a converter that we can't access or instantiate. Should we?

Reply:
Yes, I think we should.

Reply:
Issue created https://github.com/dotnet/runtime/issues/87140